### ISSUE TITLE
Horde.Registry doesn't remove keys from removed members

### ISSUE THREAD

**Issue by balena** [derekkraan/horde#203]
Consider the following supervision tree:
```
    children = [
      {Horde.Registry, keys: :unique, name: ClusterRegistry, members: :auto},
      {SomeModule, name: {:via, Horde.Registry, {ClusterRegistry, local_key}}},
      ...
    ]
```
Your `SomeModule` is just a simple `GenServer` module. Consider the following one:
```
defmodule SomeModule do
  use GenServer

  def start_link(args),
    do: GenServer.start_link(__MODULE__, nil, name: Keyword.fetch!(args, :name))

  @impl true
  def init(_), do: {:ok, []}
end
```
Say that you're using `Cluster.Strategy.Gossip` strategy from `libcluster`, but the case applies even if you don't use `libcluster`.

Start two instances side by side and connect them. Consider the `local_key` unique for each instance. While both nodes are connected, you shall see:
```
iex> Horde.Registry.select(ClusterRegistry, [{{:"$1", :"$2", :"$3"}, [], [{{:"$1", :"$2", :"$3"}}]}])
[
  {{SomeModule, "key1"}, #PID<0.480.0>, :"host1@127.0.0.1"},
  {{SomeModule, "key2"}, #PID<35396.467.0>, :"host2@127.0.0.1"}
]
```
Which is expected. Now, drop one of the two instances and from the remaining one repeat the command above:
```
iex> Horde.Registry.select(ClusterRegistry, [{{:"$1", :"$2", :"$3"}, [], [{{:"$1", :"$2", :"$3"}}]}])
[
  {{SomeModule, "key1"}, #PID<0.480.0>, :"host1@127.0.0.1"},
  {{SomeModule, "key2"}, #PID<35396.467.0>, :"host2@127.0.0.1"}
]
```
Which is not expected.

~It's interesting to note that if you pass a `listeners` parameter to `Horde.Registry`, it _does send_ a `{:unregister, registry, key, pid}` to the listener, however the key isn't removed from the registry.~ (That's not correct, a second analysis showed just `{:register, registry, key, pid, value}` coming.

In order to circumvent this problem I created a `Janitor` process to garbage collect these keys from `Horde.Registry` when a node exits the cluster. But I ask: is this behaviour "by design"? Wouldn't it `Process.monitor/1` registered processes to avoid leaving garbage in the registry forever?

**Comment by balena** [comment]
I have built a project demonstrating the problem [https://github.com/balena/horde-registry-bug](here).

**Comment by derekkraan** [comment]
Hi @balena,

This is by design. The problem is that we cannot tell the difference between a node that has exited the cluster, and a node that is inaccessible and will be rejoining the cluster later.

I suspect that finding a good solution to this problem will involve re-evaluating the CRDT that we are using to sync updates between nodes.

Derek

**Comment by balena** [comment]
Hi @derekkraan,

So if you call `Horde.Cluster.set_members` removing a member that has registered in your `Horde.Registry` the idea is to leave a dangling pid + key + value in the cluster?

The responsibility of adding/removing members isn't coming from `Horde` in the case I am mentioning. Say your cluster is dynamic, has scaled up to accommodate extra load and extra nodes made themselves visible from the cluster registry, but now they do need to leave. I don't see the point of leaving the keys in the registry in this situation: processes attempting to route jobs, or distribute load, might want to consult the registry to find which cluster processes are available, and they will always fail if any of these dangling keys are used. And moreover, you called `Horde.Cluster.set_members` removing these nodes (either your cluster strategy or your service discovery decided that these nodes are now strictly down).

In the pull request I am just handling this case: if `Horde.Cluster.set_members` remove a member, then its keys are also removed from the registry.

Guilherme.

**Comment by derekkraan** [comment]
Solved in 0779b2b07cbcce8ad58517641ab622de915ed036
